### PR TITLE
Trace improvements

### PR DESCRIPTION
**Pull request by neta-elad**

### CODE REVIEW

**Comment by def-gthill** [comment]
Thanks, I'll take a look at this on the weekend, along with @AdamantConlanger's bug reports.

**Comment by def-gthill** [comment]
For commit [e4a13fb](https://github.com/def-gthill/lexurgy/pull/63/commits/e4a13fbec130c1eea2481736a9e6114d4fa45546) (
save trace when intermediates are off): Good catch on this, yes it's almost certainly a bug. I pretty much always have intermediates on so I missed this. Good work, no changes needed!

For commit [80c7896](https://github.com/def-gthill/lexurgy/pull/63/commits/80c7896553c8fd97f3e0fc02df407fb8cf7e1ad8) (add syllables to trace): Great work again. I'm pretty sure this fixes #55, and in exactly the way I would have fixed it.

For commit [d295f61](https://github.com/def-gthill/lexurgy/pull/63/commits/d295f6188edfb0d393ca395dcb46c748bf96b59a) (show original word in trace): I can see what you're going for here, but I'm not sure I like it as is. If I'm only tracing one word, I get "Applied some-sound-change to cukucaku" over and over again. It's a bit cluttered for no gain.

If I'm tracing more than one word, I think showing the original word like this is a great idea, to make it clear which changes are being applied to which words. At least once I've used the trace feature to figure out why two similar words were diverging, and showing the original word would've helped a lot with that.

My suggestion:

- Only show the original word when there's more than one word being traced.
- Add a line to the beginning listing the words that are being traced, e.g. `tracing karapuna, teyoomi`. This provides the context when I'm only tracing one word, and also makes it clearer what's happening in situations where one of the traced words doesn't change at all.

File: src/commonMain/kotlin/com/meamoria/lexurgy/sc/SoundChanger.kt [hunk 1]
```diff
@@ -46,17 +46,18 @@ class SoundChanger(
         romanize: Boolean = true,
         debug: (String) -> Unit = ::println,
     ): Map<String?, List<String>> {
-        val debugIndices = words.withIndex().filter { it.value in debugWords }.map { it.index }
+        val debugIndices = words.withIndex().filter { it.value in debugWords }.associateBy({ it.index }, { it.value })
```
Comment (def-gthill) [inline src/commonMain/kotlin/com/meamoria/lexurgy/sc/SoundChanger.kt:49]: Please rename this variable; it's no longer just holding the indices!

Also, I think `associate { it.index to it.value }` is a bit clearer than `associateBy({ it.index }, { it.value })`.

**Review by def-gthill — CHANGES_REQUESTED** [review]

**Comment by def-gthill** [comment]
Also I really appreciate that you made sure each of your changes had a test case!

File: src/commonMain/kotlin/com/meamoria/lexurgy/sc/SoundChanger.kt (hunk 1 above)
Comment (neta-elad) [reply src/commonMain/kotlin/com/meamoria/lexurgy/sc/SoundChanger.kt:49]: I wasn't sure how to call it... How does `indexToDebugWords` sound?

Will change the `associateBy`

Comment (def-gthill) [reply src/commonMain/kotlin/com/meamoria/lexurgy/sc/SoundChanger.kt:49]: Works for me!

**Comment by neta-elad** [comment]
I fixed your comments and did a small refactor of the code that prints the trace line (to remove code duplication):
- It always prints a `Tracing ...` headline
- It only prints `to word` when there are many words
- All tracing is done in a small, new inner class `Tracer`

**Comment by def-gthill** [comment]
Thanks, looks good!

**Pull request merged**